README
Test-Device-SerialPort
======================

Test class to pretend we are working with a real device attached to serial port.
The fake object that gets created only has some methods implemented and it hasn't
the complexity of a real serial line.

For now, it does what I need...
It should be interesting to subclass it to something like
Test::Device::SerialPort::Modem or T::D::S::Gsm or something like that.
Who knows?

Lots of additional methods and some test cases added by Bill Birthisel.
It now runs under Windows in place of Win32::SerialPort as well as
under the OS's supported by Device::SerialPort (March 2010). Most of
the commonly used functions emulate as advertised.

June, 2010: Added example that syncronizes with a free-running data stream

September, 2010: Add tied filehandle support and additional examples

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

If you don't have make, Makefile.PL also creates pure perl scripts
to test and install (mostly for Win32 users).
 
SOURCE CODE REPOSITORY

  http://github.com/cosimo/perl5-test-device-serialport/


COPYRIGHT AND LICENCE

Copyright (C) 2007-2010 by Streppone Cosimo

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Test::Device::SerialPort.

0.06  Sat Sep 04 21:06:00 2010
	- Add ControlX10:xx examples and read_back_cmd
	- Add write_decoder capability and removed unused cm11_write
	- Add tied filehandle support and demo
	- Improved documentation

0.05  Thu Jun 10 12:59:04 2010
	- Add status and is_status methods
	- Work around POSIX.pm warnings on Windows

0.04  Tue Jun 01 15:55:00 2010 (Bill Birthisel)
	- Add demo eg/elect_meter.pl
	- Allow 1200 baud
	- resolve timing test failures

0.03  Tue Mar 23 15:15:15 2010
	- Add many methods from Win32 and Device versions (major rewrite)
	- Add tests
	- Add Win32 and ControlX10 compatibility
	- changes by Bill Birthisel

0.02  Wed Mar 03 03:03:03 2010 (time estimated)
	- create github repository

0.01  Wed Feb 14 15:07:07 2007
	- original version; created by h2xs 1.23 with options
		-A -X -b 5.6.1 -n Test::Device::SerialPort



t/AltPort.pm
package AltPort;
# Inheritance test for test3.t only

our $VERSION = '0.06';
require Exporter;
use Test::Device::SerialPort qw( :STAT :PARAM 0.06 );
our @ISA = qw( Exporter Test::Device::SerialPort );
our @EXPORT= qw();
our @EXPORT_OK= @Test::Device::SerialPort::EXPORT_OK;
our %EXPORT_TAGS = %Test::Device::SerialPort::EXPORT_TAGS;

my $in = BM_fCtsHold;
print "AltPort import=$in\n";
1;


Makefile.PL
use strict;
use warnings;
my $version = "0.06";	# for nomake versions
use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR            => 'Cosimo Streppone <cosimo@cpan.org>',
    NAME              => 'Test::Device::SerialPort',
    LICENSE           => 'perl',
    ABSTRACT_FROM     => 'lib/Test/Device/SerialPort.pm',
    VERSION_FROM      => 'lib/Test/Device/SerialPort.pm',
    PREREQ_PM         => {
		'Test::More' => 0,
	},
    META_MERGE => {
        resources => {
            repository => 'git://github.com/cosimo/perl5-test-device-serialport.git',
            bugtracker => 'mailto:bug-test-device-serialport@rt.cpan.org',
            license    => 'http://dev.perl.org/licenses/',
        },
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Test-Device-SerialPort-* *.cfg t/*.cfg' },
    realclean       => { 'FILES' => "nomake_test nomake_install" },
);

print <<COMMENTS;
            Test::Device::SerialPort VERSION $version

A 'Makefile' is created for those with 'make' or CPAN.pm users.
The normal 'Mantra' would then apply:
        make
        make test
        make install

For those without 'make' or an equivalent like 'nmake' or 'dmake' there
are perl-only scripts which do the same things:
        Test with:    perl nomake_test
        Install with: perl nomake_install

Timeout tests can take up to 30 seconds per test.
COMMENTS

my $dfile = "nomake_test";
unlink $dfile;
print "Creating new $dfile\n";
open (DEFAULT, "> $dfile") or die "Can't create $dfile: $!\n";

print DEFAULT <<TEST4;
# Created by Makefile.PL
# VERSION $version

use Test::Harness;
runtests ("t/Test-Device-SerialPort.t", "t/test2.t", "t/test3.t");

# See README and other documentation for additional information.
TEST4

close DEFAULT;

$dfile = "nomake_install";
unlink $dfile;
print "Creating new $dfile\n";
open (DEFAULT, "> $dfile") or die "Can't create $dfile: $!\n";

print DEFAULT <<INST4;
# Created by Makefile.PL
# VERSION $version

use 5.006;

use Config;
use strict;
use warnings;
use ExtUtils::Install qw( install );

my \$FULLEXT = "Test/Device/SerialPort";
my \$INST_LIB = "./lib";
install({
	   read => "\$Config{sitearchexp}/auto/\$FULLEXT/.packlist",
	   write => "\$Config{installsitearch}/auto/\$FULLEXT/.packlist",
	   \$INST_LIB => "\$Config{installsitelib}",
	  },1,0,0);

__END__
INST4

close DEFAULT;



eg/tie_demo.pl
use lib './lib','../lib';	# can run before final install

## This implements the same tied actions as demo6 in the
## Win32 and Device::SerialPort distributions
## But simulates the results

use Test::Device::SerialPort 0.06;
require 5.008;

use strict;
use warnings;

package main;	# default, but safe to know when using write_decoder

# loopback responses
sub test_write_decoder {
    return unless (@_ == 2);
    my $self = shift;
    my $wbuf = shift;
    my $response = "";
    return unless ($wbuf);
    if ($wbuf =~ /one character/) {
	$response = 'A';
    } elsif ($wbuf =~ /enter line/) {
	$response = "The quick brown fox jumped over the lazy dog\n";
    } elsif ($wbuf =~ /enter 5 char/) {
	$response = "Abcde";
    } elsif ($wbuf =~ /enter 5 more/) {
	$response = "Fghij";
    } else {
	$response = 'Not Found';
    }
    $self->lookclear($response);
    my $sent = $wbuf;
    $sent =~ s/^\s*//omg;	# strip leading whitespace
    warn "RECEIVED: $sent\n";	# see what was sent
    return length($wbuf);
}

my $cfgfile = "DEMO.cfg";
unlink $cfgfile;

my $ob = Test::Device::SerialPort->new('DEMO');
die "Can't open serial port DEMO\n" unless ($ob);

$ob->save($cfgfile);
undef $ob;

my $head	= "\r\n\r\n+++++++++++ Tied FileHandle Demo ++++++++++\r\n";
my $e="\r\n....Bye\r\n";

    # constructor = TIEHANDLE method
my $tie_ob = tie(*PORT,'Test::Device::SerialPort', $cfgfile)
                 || die "Can't start $cfgfile\n";

    # match parameters
$tie_ob->are_match("\n");
$tie_ob->set_no_random_data(1);
$tie_ob->lookclear;

    # setup loopback write decoding
$tie_ob->write_decoder('main::test_write_decoder');

    # Print Prompt to Port
print PORT $head;

    # tie to PRINT method
print PORT "\r\nEnter one character (10 seconds): "
    or die "PRINT timed out\n";	# should never happen in loopback

    # tie to GETC method
my $char = getc PORT;
print PORT "$char\r\n";

    # tie to WRITE method
my $out = "\r\nThis is a 'syswrite' test\r\n\r\n";
syswrite PORT, $out, length($out), 0
    or die "WRITE timed out\n";

    # tie to READLINE method
print PORT "enter line: ";
my $line = <PORT>;
print PORT "\r\nREADLINE received: $line\r";

    # tie to READ method
my $in = "FIRST:12345, SECOND:67890, END";
print PORT "\r\nenter 5 char: ";
unless (defined sysread (PORT, $in, 5, 6)) {
    die "...READ timed out\n";
}
print PORT "\r\nenter 5 more char: ";
unless (defined sysread (PORT, $in, 5, 20)) {
    die "...second READ timed out\n";
}

    # tie to PRINTF method
printf PORT "\r\nreceived: %s\r\n", $in
    or die "PRINTF timed out\n";

    # PORT-specific versions of the $, and $\ variables
my $n1 = ".number1_";
my $n2 = ".number2_";
my $n3 = ".number3_";

print PORT $n1, $n2, $n3;
print PORT "\r\n";

$tie_ob->output_field_separator("COMMA");
print PORT $n1, $n2, $n3;
print PORT "\r\n";

$tie_ob->output_record_separator("RECORD");
print PORT $n1, $n2, $n3;
$tie_ob->output_record_separator("");
print PORT "\r\n";
    # the $, and $\ variables will also work

print PORT $e;

    # destructor = CLOSE method
close PORT || print "CLOSE failed\n\n";

    # destructor = DESTROY method
undef $tie_ob;	# Don't forget this one!!
untie *PORT;


eg/elec_meter.pl
#!/usr/bin/perl
#---------------------------------------------------------------------------
#  Title:
#      Cross-Platform Demo - "uses" right module on either Win32 or linux
#  Usage:
#      perl elec_meter.pl PORT
#  if PORT eq 'TEST' uses Test::Device:SerialPort emulator
#---------------------------------------------------------------------------

# must be LF-only line ends to run on all platforms

use lib './lib','../lib';	# can run before final install

use strict;
use warnings;
use Data::Dumper;

our $OS_win;
our $ob;
our $port;

BEGIN {
	die "\nUsage: perl elec_meter.pl PORT\n" unless (@ARGV);
	$port = shift @ARGV;
	my $eval_module;
	my $module_version;

        $OS_win = ($^O eq "MSWin32" || $^O eq "cygwin") ? 1 : 0;

        print "Perl version: $]\n";
        print "OS   version: $^O\n";

            # This must be in a BEGIN in order for the 'use' to be conditional
        if ($port eq 'TEST') {
        	print "Loading Emulator module\n";
        	$eval_module = "Test::Device::SerialPort";
		$module_version = 0.04;
        }
        elsif ($OS_win) {
        	print "Loading Windows module\n";
        	$eval_module = "Win32::SerialPort";
		$module_version = 0.21;
        }
        else {
            print "Loading Unix module\n";
            $eval_module = "Device::SerialPort";
	    $module_version = 1.04;
        }
        my $eval_str = "use $eval_module qw( :STAT $module_version ); \$ob = $eval_module->new('$port');";
	warn $eval_str . "\n";
	eval $eval_str;
	die "$@\n" if ($@);
	die "Can't open serial port $port: $^E\n" unless ($ob);
} # End BEGIN

$ob->user_msg(1); # misc. warnings
## $ob->error_msg(1); # hardware and data errors

$ob->baudrate(1200) || die "fail setting baud";
$ob->parity("even") || die "fail setting parity";
$ob->parity_enable(1);
$ob->databits(7) || die "fail setting databits";
$ob->stopbits(1) || die "fail setting stopbits";
$ob->handshake("none") || die "fail setting handshake";

$ob->write_settings || die "no settings";

$ob->are_match("\cM\cC");	# end string = CR ETX

if ($port eq 'TEST') {
	# emulate meter output
	$ob->set_test_mode_active(1);
	my $data = "PAPP 00400 %\cM\cJ";	# start partway thru pattern
	$data .= "MOTDETAT 000000 B\cM\cC\cB\cJ";
	$data .= "ADCO 020322590713 9\cM\cJ";
	$data .= "OPTARIF BASE 0\cM\cJ";
	$data .= "ISOUSC 30 9\cM\cJ";
	$data .= "BASE 024576277 3\cM\cJ";
	$data .= "PTEC TH.. \$\cM\cJ";
	$data .= "IINST 002 Y\cM\cJ";
	$data .= "IMAX 026 G\cM\cJ";
	$data .= $data;
	$data .= $data;		# total of 4 sets of pattern
	$ob->lookclear($data);		# preset buffers
} else {
	# really read from the meter
	$ob->lookclear;		# empty buffers
}
my $gotit = "";
my $match1 = "";
until ("" ne $gotit) {
	if ($OS_win) {
		# *ix handles errors differently
		my @stat = $ob->status;
		if ($stat[ST_ERROR]) {
			$ob->reset_error;
		}
	}
	$gotit = $ob->streamline;	# poll until data ready
	last if ($gotit);
	$match1 = $ob->matchclear;	# match is first thing received
	last if ($match1);
	sleep 1;
}

# so to get here, we have seen an ETX aned synced up
# but we don't know how much of a transmission was caught
# so we discard it and start new_but_sync'd

$gotit = "";
until ("" ne $gotit) {
	# I doubt if we need the status check here, but it does no harm
	# reset_error only needed on Windows
	if ($OS_win) {
		my @stat = $ob->status;
		if ($stat[ST_ERROR]) {
			$ob->reset_error;
		}
	}
	$gotit = $ob->streamline;	# poll until next ETX
}

# so let's see what we actually got

$gotit =~ s/\cB\cJ//g;	# remove STX LF from start

my @readings = split ("\cM\cJ", $gotit);
push @readings, 'BAD_ONE 00610 (';	# added for test, WCB
## warn Dumper \@readings;

my %results;

foreach my $r (@readings) {
	my $sum = 0;
	my $csum = chop $r; # remove and save checksum
	chop $r; # remove space before checksum
	my @char = unpack ('C*', $r);
	foreach my $c (@char) {
		$sum += $c;
	}
	# print "$r\n@char\n";
	my $sum2 = $sum & 0x3f;
	$sum2 += 0x20;
	my $csum2 = chr($sum2);
	# printf "sum=%x, sum2=%x, %s..%s\n", $sum, $sum2, $csum2, $csum;
	my ($item, $value) = split (' ', $r);
	if ($csum eq $csum2) {
		$results{$item} = $value;
	} else {
		$results{$item} = 'invalid';
	}
}
warn Dumper \%results;

$ob->close || die "\nclose problem with $port\n";
undef $ob;
